star_realms/src/game/effects/mod.rs

```diff
@@ -109,6 +112,64 @@ pub fn is_free_cond(cond: &String) -> bool {
     }
 }
 
+impl PreConfig {
+    pub fn create(map: HashMap<String, String>) -> PreConfig {
+        PreConfig {
+            map,
+            validated: false
+        }
+    }
+    pub fn validate (&self, meta: &PreConfigMeta) -> Failure<String> {
+        for k in meta.required_keys.iter() {
+            if !self.map.contains_key(k) {
+                return Fail(format!("Requires the key '{}'", k));
+            }
+        }
+        for (k, s) in self.map.iter() {
+            if let Some(t) = meta.types.get(k) {
+                return match t {
+                    PreConfigType::Nat => try_parse_or::<u32>(
+                        s,
+                        format!(">>{}: {}<< must be of the Nat type", k, s)
+                    ),
+                    PreConfigType::String => try_parse_or::<u32>(
+                        s,
+                        format!(">>{}: {}<< must be of the String type", k, s)
+                    )
+                }
+            }
+        }
+        Succeed
+    }
+    pub fn get_nat(&self, key: &str) -> u32 {
+        self.map.get(key).unwrap().parse().unwrap()
+    }
+    pub fn get_str(&self, key: &str) -> &str {
+        self.map.get(key).unwrap()
+    }
+}
+
+impl PreConfigMeta {
+    fn all_required(kvs: Vec<(&str, PreConfigType)>) -> PreConfigMeta {
```

Comment:
aliases would be nice but not sure how to implement those
maybe communicate through methods rather than the public data in `PreConfigMeta`

Reply:
with an `is_key` method